File: c++/226.invert-binary-tree.cpp
/*
 * @lc app=leetcode id=226 lang=cpp
 *
 * [226] Invert Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <stack>

// Accepted
// 77/77 cases passed (8 ms)
// Your runtime beats 12.36 % of cpp submissions
// Your memory usage beats 5.93 % of cpp submissions (9.8 MB)
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
		if (root == nullptr)
			return root;
		invertTree(root -> left);
		invertTree(root -> right);
		TreeNode* tempL = root -> left;
		root -> left = root -> right;
		root -> right = tempL;
        return root;
    }
};
// @lc code=end

File: python/226.invert-binary-tree.py
#
# @lc app=leetcode id=226 lang=python3
#
# [226] Invert Binary Tree
#

# Runtime: 46 ms, faster than 49.74% of Python3 online submissions for Invert Binary Tree.
# Memory Usage: 13.7 MB, less than 95.84% of Python3 online submissions for Invert Binary Tree.

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


# My solution:
# class Solution:
#     def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
#         if not root:
#             return None

#         temp = root.left
#         root.left = root.right
#         root.right = temp

#         self.invertTree(root=root.left)
#         self.invertTree(root=root.right)

#         return root
        
# Solution from leet code for studying:
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root:
            root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)
            return root

# @lc code=end


class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        stack = [root]
        while stack:
            node = stack.pop()
            if node:
                node.left, node.right = node.right, node.left
                stack.extend([node.right, node.left])
        return root
